File: Java/src/civitas/Tablero.java
package civitas;

import java.util.ArrayList;

public class Tablero {  
  
  private int numCasillaCarcel;
  private int porSalida;
  private Boolean tieneJuez;
  private ArrayList<Casilla> casillas;
  
  void añadeCasilla (Casilla casilla) {
    if (casillas.size() == numCasillaCarcel) {
      Casilla carcel = new Casilla("Cárcel");
      casillas.add(carcel);
    }
    casillas.add(casilla);
    if (casillas.size() == numCasillaCarcel) {
      Casilla carcel = new Casilla("Cárcel");
      casillas.add(carcel);
    }
  }

  void añadeJuez () {
    if (!tieneJuez) {
      // System.err.println("Estoy en tablero.añadeJuez. Carcel en "+getCarcel());
      Casilla juez = new Casilla(getCarcel(), "Juez");
      casillas.add(juez);
      tieneJuez = true;
    }
  }
  
  int calcularTirada(int origen, int destino) {
    int posicion = destino - origen;
    if (posicion < 0) 
    posicion += casillas.size();
    return posicion;
  }
  
  private Boolean correcto () {
    Boolean valido = false;
    if (casillas.size() > numCasillaCarcel && tieneJuez)
    valido = true;
    return valido;
  }
  
  private Boolean correcto (int numCasilla) {
    Boolean valido = false;
    if (correcto() && numCasilla <= casillas.size() && numCasilla >= 0)
      valido = true;
    return valido;
  }
  
  int getCarcel () {
    return numCasillaCarcel;
  }
  
  Casilla getCasilla (int numCasilla) {
    Casilla casilla = null;
    if (correcto(numCasilla))
      casilla = casillas.get(numCasilla);
    return casilla;
  } 
  
  public ArrayList<Casilla> getCasillas() {
    return casillas;
  }

  int getPorSalida () {
    int ret = porSalida;
    if (porSalida > 0)
      porSalida--;
    return ret;
  }

  int nuevaPosicion (int actual, int tirada) {
    int posicion = -1;
    if (correcto()) {
      posicion = (actual + tirada) % casillas.size();
      if (posicion != (actual + tirada)) 
        porSalida++;
    }
    return posicion;
  }

  public Tablero (int carcel) {
    numCasillaCarcel = 1;
    if (carcel>0)
    numCasillaCarcel = carcel;
    casillas = new ArrayList<>();
    Casilla Salida = new Casilla("Salida");
    casillas.add(0, Salida);
    porSalida = 0;
    tieneJuez = false;
  }

}

File: Ruby/lib/civitas-ruby/tablero.rb
require_relative 'casilla'

module Civitas
  class Tablero
    
    def initialize(num_casilla_carcel)
      @num_casilla_carcel = num_casilla_carcel >= 1 ? num_casilla_carcel : 1
      @casillas           = [Casilla.new_descanso("Salida")]
      @por_salida         = 0
      @tiene_juez         = false
    end
    
    attr_reader :num_casilla_carcel, :casillas

    def por_salida
      ret = @por_salida
      @por_salida -=1 if @por_salida > 0
      ret
    end

    def añade_casilla(casilla)
      @casillas << Casilla.new_descanso("Cárcel") if @casillas.size == @num_casilla_carcel
      @casillas << casilla
      @casillas << Casilla.new_descanso("Cárcel") if @casillas.size == @num_casilla_carcel
    end

    def añade_juez
      if !@tiene_juez
        @casillas << Casilla.new_juez(@num_casilla_carcel, "Juez")
        @tiene_juez = true
      end
    end

    def nueva_posicion(actual, tirada)
      result = 
        if !correcto
          -1
        else 
          posicion = (actual + tirada) % @casillas.size
          @por_salida += 1 if posicion != actual + tirada
          posicion
        end
    end

    def calcular_tirada(origen, destino)
      result = destino - origen
      result += @casillas.size if result < 0
      result
    end

    private

    def correcto(num_casilla = 0)
      (@casillas.size > @num_casilla_carcel) && @tiene_juez && (num_casilla < @casillas.size)  
    end

  end
end
